### ISSUE TITLE
Scheduler violates hard anti-affinity constraint when sidekick present

### ISSUE THREAD

**Issue by LLParse** [rancher/rancher#8070]
**Rancher Versions:**
Server: v1.4.1
scheduler: v0.6.3

**Docker Version:** 1.12.6

**OS and where are the hosts located? (cloud, bare metal, etc):** AWS

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)** single internal

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)** cattle

**Steps to Reproduce:**
1. Add 7 hosts to a cattle environment
2. Create a 7-node etcd from community catalog
3. Power cycle 3 hosts

**Results:** 
2 etcd land on same host, violating `io.rancher.scheduler.affinity:container_label_ne: io.rancher.stack_service.name=$${stack_name}/$${service_name}` constraint

![image](http://i.imgur.com/ydh2lm5.png)

EDIT: Removing the sidekick data container appears to have fixed this, so it is a distinct possibility that only services with sidekicks are afflicted.

**Comment by LLParse** [comment]
This is particularly dangerous because these 2 containers end up consuming the same named volume, potentially corrupting the data.

**Comment by LLParse** [comment]
I just reproduced this with a _fresh_ deployment of etcd, which leads me to believe this is a regression.

![fresh install](http://i.imgur.com/ylLR401.png)

**Comment by schuylr** [comment]
Can confirm the issues above. I have a theory concerning the data containers as I was running a modified version of your latest etcd image without them, I'm going to try a modified service and see if it helps.

**Comment by schuylr** [comment]
@LLParse Here's the docker and rancher compose that I'm running as a modified service. Basically I have removed the data sidekick containers.

docker-compose.yml:
```
etcd:
    image: rancher/etcd:v2.3.7-11
    labels:
        io.rancher.scheduler.affinity:host_label_soft: etcd=true
        io.rancher.scheduler.affinity:container_label_ne: io.rancher.stack_service.name=$${stack_name}/$${service_name}
    environment:
        RANCHER_DEBUG: 'true'
        EMBEDDED_BACKUPS: 'true'
        BACKUP_PERIOD: '15m'
        BACKUP_RETENTION: '24h'
    volumes:
    - /var/etcd-test:/data-backup
```

rancher-compose.yml:
```
etcd:
  scale: 7
  retain_ip: true
  health_check:
    port: 2378
    request_line: /health
    interval: 5000
    response_timeout: 3000
    unhealthy_threshold: 2
    healthy_threshold: 2
```

I don't see any affinity issues yet. I think the data containers (or something with the way Cattle handles sidekick or started-once containers) triggers the problem but I'm not sure. Since you have a much quicker way to reproduce the problem (I don't have 7 test machines spun up right now) - would you like to try this one out?

**Comment by LLParse** [comment]
@schuylr Thanks. I want to get rid of that data container anyways, it serves no purpose beyond upgrading from an old ephemeral template. Removing it would only make the scheduling logic simpler, but should work regardless.

I'll give this one a shot real quick and see what happens.

**Comment by LLParse** [comment]
Tested without persistence/data containers and scheduling worked flawlessly. Cycled 3 hosts and observed successful reconciliation.

**Comment by schuylr** [comment]
Excellent! I'm glad I was able to help hone in on reproducing the issue.

I can also confirm that #8066 also doesn't happen with me now thanks to the lack of data containers.

So my workaround for now will be to use the above compose configurations. I'll run the risk of not having persistence since there are backups anyways.

**Comment by schuylr** [comment]
Here's another interesting point, I don't see the issue with the community-supported Zookeeper service, which also has data sidekicks. I was trying this service out to compare its reliability to the etcd catalog entry.

![screen shot 2017-03-06 at 3 13 20 pm](https://cloud.githubusercontent.com/assets/12628715/23627626/7656af78-027f-11e7-83e3-4f3011c39e6c.png)

The service logs are empty, I've left this service running all weekend, and all containers are still healthy with no issues.

The only difference with this entry is that there is no soft affinity rule, only the hard anti-affinity rule. I'm going to add the soft affinity rule of `etcd=true` to the Zookeeper cluster and see if it Rancher screws it up.

**Comment by LLParse** [comment]
Perhaps the mix of hard container anti-affinity with soft host affinity _with a sidekick_ is the cause. I won't know until somebody hooks up a debugger, but we can reliably reproduce this now so it shouldn't be difficult for the devs to figure out.

I really don't know what to recommend at this point, but will be interested to hear how things fare with the modified compose. For reference, we run almost the same etcd in our kubernetes stack at a hard scale limit of 3, and haven't seen this particular thing happen.

**Comment by schuylr** [comment]
It has been working with my environment for an hour now. If the modified compose keeps working for me, I probably won't have much to report, but will report it here. No news is good news 😃

**Comment by ryanwalls** [comment]
@schuylr any updates on your setup?  Has it still been working for you without sidekick containers?

**Comment by schuylr** [comment]
@ryanwalls Absolutely flawlessly. I haven't had a single etcd issue since my last comment here, knock on wood.

**Comment by sangeethah** [comment]
I see this behavior where etc services land on the host where other etc instances are already running , when upgrading etcd services using rancher-cli command  ```up --force-upgrade -d```  with batch size set to 1 for k8s stack.
etcd is not able to successfully upgrade after this.


<img width="1180" alt="screen shot 2017-04-24 at 4 56 58 pm" src="https://cloud.githubusercontent.com/assets/4266958/25363465/7ad77724-290f-11e7-9d62-606b78a2369b.png">


```
4/24/2017 4:56:19 PM+ META_URL=http://rancher-metadata.rancher.internal/2015-12-19
4/24/2017 4:56:19 PM++ wget -q -O - http://rancher-metadata.rancher.internal/2015-12-19/self/stack/name
4/24/2017 4:56:19 PM+ STACK_NAME=kubernetes
4/24/2017 4:56:19 PM+ '[' kubernetes == '' ']'
4/24/2017 4:56:19 PM++ giddyup service scale etcd
4/24/2017 4:56:19 PM+ SCALE=3
4/24/2017 4:56:19 PM++ giddyup ip myip
4/24/2017 4:56:19 PM+ IP=10.42.55.201
4/24/2017 4:56:19 PM++ wget -q -O - http://rancher-metadata.rancher.internal/2015-12-19/self/container/create_index
4/24/2017 4:56:19 PM+ CREATE_INDEX=13
4/24/2017 4:56:19 PM++ wget -q -O - http://rancher-metadata.rancher.internal/2015-12-19/self/container/service_index
4/24/2017 4:56:19 PM+ SERVICE_INDEX=2
4/24/2017 4:56:19 PM++ wget -q -O - http://rancher-metadata.rancher.internal/2015-12-19/self/host/uuid
4/24/2017 4:56:19 PM+ HOST_UUID=ae1d8f0a-c928-4203-a22d-145fa76d7fac
4/24/2017 4:56:19 PM+ LEGACY_DATA_DIR=/data
4/24/2017 4:56:19 PM+ DATA_DIR=/pdata
4/24/2017 4:56:19 PM+ DR_FLAG=/pdata/DR
4/24/2017 4:56:19 PM+ export ETCD_DATA_DIR=/pdata/data.current
4/24/2017 4:56:19 PM+ ETCD_DATA_DIR=/pdata/data.current
4/24/2017 4:56:19 PM+ export ETCDCTL_ENDPOINT=http://etcd.kubernetes:2379
4/24/2017 4:56:19 PM+ ETCDCTL_ENDPOINT=http://etcd.kubernetes:2379
4/24/2017 4:56:19 PM++ echo 10.42.55.201
4/24/2017 4:56:19 PM++ tr . -
4/24/2017 4:56:19 PM+ NAME=10-42-55-201
4/24/2017 4:56:19 PM+ '[' 1 -eq 0 ']'
4/24/2017 4:56:19 PM+ eval node
4/24/2017 4:56:19 PM++ node
4/24/2017 4:56:19 PM++ mkdir -p /pdata/data.current
4/24/2017 4:56:19 PM++ '[' -d /data/member ']'
4/24/2017 4:56:19 PM++ '[' -d /data/data.current ']'
4/24/2017 4:56:19 PM++ '[' -f /pdata/DR ']'
4/24/2017 4:56:19 PM++ '[' -d /pdata/data.current/member ']'
4/24/2017 4:56:19 PM+++ cat /pdata/data.current/ip
4/24/2017 4:56:19 PM++ '[' 10.42.55.201 == 10.42.55.201 ']'
4/24/2017 4:56:19 PM++ restart_node
4/24/2017 4:56:19 PM++ healthcheck_proxy
4/24/2017 4:56:19 PM++ WAIT=60s
4/24/2017 4:56:19 PM++ etcdwrapper healthcheck-proxy ++ etcd --port=:2378 --name 10-42-55-201 --listen-client-urls http://0.0.0.0:2379 --advertise-client-urls http://10.42.55.201:2379 --listen-peer-urls --wait=60s --debug=false
4/24/2017 4:56:19 PMhttp://0.0.0.0:2380 --initial-advertise-peer-urls http://10.42.55.201:2380 --initial-cluster-state existing
4/24/2017 4:56:19 PM++ rolling_backup
4/24/2017 4:56:19 PM++ EMBEDDED_BACKUPS=true
4/24/2017 4:56:19 PM++ '[' true == true ']'
4/24/2017 4:56:19 PM++ BACKUP_PERIOD=15m0s
4/24/2017 4:56:19 PM++ BACKUP_RETENTION=24h
4/24/2017 4:56:19 PM++ giddyup leader elect --proxy-tcp-port=2160 etcdwrapper rolling-backup --period=15m0s --retention=24h --index=2
4/24/2017 4:56:19 PMtime="2017-04-24T23:56:19Z" level=info msg="Listening on 0.0.0.0:2160"
4/24/2017 4:56:19 PMtime="2017-04-24T23:56:19Z" level=info msg="Forwarding setup to: :2160"
4/24/2017 4:56:19 PM2017-04-24 23:56:19.395254 I | flags: recognized and used environment variable ETCD_DATA_DIR=/pdata/data.current
4/24/2017 4:56:19 PM2017-04-24 23:56:19.395535 I | flags: recognized and used environment variable ETCD_ELECTION_TIMEOUT=5000
4/24/2017 4:56:19 PM2017-04-24 23:56:19.395625 I | flags: recognized and used environment variable ETCD_HEARTBEAT_INTERVAL=500
4/24/2017 4:56:19 PM2017-04-24 23:56:19.395962 I | etcdmain: etcd Version: 2.3.7
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396026 I | etcdmain: Git SHA: fd17c91
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396103 I | etcdmain: Go Version: go1.6.2
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396162 I | etcdmain: Go OS/Arch: linux/amd64
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396229 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396493 W | etcdmain: found invalid file/dir ip under data dir /pdata/data.current (Ignore this if you are upgrading etcd)
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396565 N | etcdmain: the server is already initialized as member before, starting as etcd member...
4/24/2017 4:56:19 PM2017-04-24 23:56:19.396879 I | etcdmain: listening for peers on http://0.0.0.0:2380
4/24/2017 4:56:19 PM2017-04-24 23:56:19.397084 I | etcdmain: listening for client requests on http://0.0.0.0:2379
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398414 I | etcdserver: name = 10-42-55-201
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398497 I | etcdserver: data dir = /pdata/data.current
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398560 I | etcdserver: member dir = /pdata/data.current/member
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398600 I | etcdserver: heartbeat = 500ms
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398672 I | etcdserver: election = 5000ms
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398720 I | etcdserver: snapshot count = 10000
4/24/2017 4:56:19 PM2017-04-24 23:56:19.398806 I | etcdserver: advertise client URLs = http://10.42.55.201:2379
4/24/2017 4:56:19 PM2017-04-24 23:56:19.399319 C | etcdserver: open wal error: file already locked
4/24/2017 4:56:19 PM++ cleanup 1
4/24/2017 4:56:19 PM++ exitcode=1
4/24/2017 4:56:19 PM+++ date -R
4/24/2017 4:56:19 PM++ timestamp='Mon, 24 Apr 2017 23:56:19 +0000'
4/24/2017 4:56:19 PM++ echo 'Exited (1)'
4/24/2017 4:56:19 PMExited (1)
4/24/2017 4:56:19 PM++ '[' 1 == 0 ']'
4/24/2017 4:56:19 PM++ '[' 1 == 2 ']'
4/24/2017 4:56:19 PM++ echo 'Mon, 24 Apr 2017 23:56:19 +0000 -> Exit (1), unknown. No action taken'
4/24/2017 4:56:19 PM++ exit 1
```

**Comment by cjellick** [comment]
Tested with same setup specified in bug: 7 node cluster and 7 node etc stack. When I rebooted hosts, etcd containers did not deploy anywhere else. Once the hosts were back up and running, they redeployed to those hosts properly. When I deleted a host (making a situation impossible to schedule because there were more expected etcd nodes than hosts), cattle correctly failed to schedule the etcd container.

Version - v1.6.0-rc3
Verified fix
